validate-index.cpp

```diff
@@ -0,0 +1,92 @@
+#include "clang/Index/IndexDataStore.h"
+#include "clang/Index/IndexUnitReader.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+
+#include <string>
+#include <utility>
+#include <vector>
+
+using namespace llvm;
+using namespace llvm::sys;
+using namespace clang::index;
+
+static cl::opt<std::string> IndexStore(cl::Positional, cl::Required,
+                                       cl::desc("<indexstore>"));
+
+int main(int argc, char **argv) {
+  cl::ParseCommandLineOptions(argc, argv);
+
+  std::string storeError{};
+  auto store = IndexDataStore::create(IndexStore, storeError);
+  if (not store) {
+    errs() << "error: failed to open indexstore " << IndexStore << " -- "
+           << storeError << "\n";
+    return EXIT_FAILURE;
+  }
+
+  std::vector<std::string> unitNames{};
+  store->foreachUnitName(false, [&](StringRef unitName) {
+    unitNames.push_back(unitName.str());
+    return true;
+  });
+
+  auto exitStatus = EXIT_SUCCESS;
+  for (const auto &unitName : unitNames) {
+    std::string readerError;
+    auto reader = IndexUnitReader::createWithUnitFilename(unitName, IndexStore,
+                                                          readerError);
+    if (not reader) {
+      errs() << "error: failed to read unit file " << unitName << " -- "
+             << readerError << "\n";
+      return EXIT_FAILURE;
```

Comment:
should this failure is cause early termination? what about other unit files?

Reply:
good call. It should never happen, but I guess not worth exiting.

Reply:
Fixed